### PR TITLE
Io/pandas

### LINKED ISSUE #47: Update IO with Pandas

### PR DESCRIPTION
**Pull request by Jacob-Chmura**
Close #47

### CODE REVIEW

File: opendg/graph.py [hunk 1]
```diff
@@ -45,6 +47,39 @@ def from_csv(
...
+            dst_col (str): Column in the dataframe corresponding to the destination edge id.
+            time_col (str): Column in the dataframe corresponding to the timestamp.
+            edge_feature_col (Optional[str]): Optional column in the dataframe corresponding to the edge features.
+            time_delta (Optional[TimeDeltaDG]): Describes the time granularity associated with the event stream.
+                If None, then the events are assumed 'ordered', with no specific time unit.
+
+        Returns:
+           DGraph: The newly constructed dynamic graph.
+        """
+        non_ordered_time_delta = time_delta is not None and not time_delta.is_ordered
```
Comment (Jacob-Chmura) [inline opendg/graph.py:74]:
```suggestion
        ordered_time_delta = time_delta is None or time_delta.is_ordered
```
much simpler to read

**Review by Jacob-Chmura — COMMENTED** [review]
Here is a quick implementation.

I think we should think carefully whether we want to include pandas IO, as it is something we will need to maintain and support, There are most likely corner cases that are not handled here, but would come up. 

I have added `pandas` as a hard package dependency for the moment. We could lazy import this (as we plan to do with `from_tgb`) though, the difference is that for pandas `DGraph` actually expects a dataframe object (instead of a string). We can still technically get away with lazy importing by skipping the type check. But at the end of the day, our library _is_ using our own pandas package in `_io/pandas.py` which makes it a hard dep.

File: opendg/_io/pandas.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+from opendg.events import EdgeEvent, Event
+
+
+def read_pandas(
+    df: pd.DataFrame,
+    src_col: str,
+    dst_col: str,
+    time_col: Optional[str] = None,
+    edge_feature_col: Optional[str] = None,
```
Comment (Jacob-Chmura) [inline opendg/_io/pandas.py:14]: These argument names make the most sense to me.

Should update them accordingly in `from_csv()`: https://github.com/shenyangHuang/openDG/blob/37fe42913ebaf067711f25e19f036e79dfc9d669/opendg/_io/csv.py#L10-L14

File: opendg/graph.py [hunk 3]
```diff
@@ -45,6 +47,39 @@ def from_csv(
...
+        ordered_time_delta = time_delta is None or time_delta.is_ordered
+        if time_col is None and not ordered_time_delta:
+            raise ValueError(
+                'Must specify "time_col" when using non-ordered time delta.'
+            )
+
+        events = read_pandas(df, src_col, dst_col, time_col, edge_feature_col)
+        return cls(events, time_delta)
+
     def to_csv(self, file_path: str, *args: Any, **kwargs: Any) -> None:
```
Comment (Jacob-Chmura) [inline opendg/graph.py:83]: Should expose the actual arguments here (as for `from_pandas()`

File: opendg/_io/pandas.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+from typing import List, Optional
+
+import numpy as np
+import pandas as pd
+import torch
+
+from opendg.events import EdgeEvent, Event
+
+
+def read_pandas(
```
Comment (shenyangHuang) [inline opendg/_io/pandas.py:10]: looks good. We should discuss the pandas dependency

File: opendg/graph.py [hunk 5]
```diff
@@ -26,23 +28,65 @@ def __init__(
...
                 If None, then the events are assumed 'ordered', with no specific time unit.
-            args (Any): Optional positional arguments.
-            kwargs (Any): Optional keyword arguments.
 
         Returns:
            DGraph: The newly constructed dynamic graph.
         """
-        events = read_csv(file_path, *args, **kwargs)
+        ordered_time_delta = time_delta is None or time_delta.is_ordered
+        if time_col is None and not ordered_time_delta:
```
Comment (shenyangHuang) [inline opendg/graph.py:52]: ordered time delta you also need to assign timestamp, it is just that sometimes you don't know what time granularity it is but you should still have timestamps.

Comment (shenyangHuang) [reply opendg/graph.py:52]: thus I would make `time_col` mandatory

File: test/test_io/test_csv_io.py [hunk 6]
```diff
@@ -71,21 +44,25 @@ def test_csv_conversion_with_features_no_feature_cols_provided():
...
     ]
+    col_names = {'src_col': 'src_id', 'dst_col': 'dst_id', 'time_col': 'time'}
 
     with tempfile.NamedTemporaryFile() as f:
         with pytest.raises(ValueError):
-            _ = write_csv(events, f.name)
+            _ = write_csv(events, f.name, **col_names)
 
 
 def test_csv_conversion_with_features_bad_feature_col_shape():
```
Comment (shenyangHuang) [inline test/test_io/test_csv_io.py:54]: what does this function do?

**Review by shenyangHuang — COMMENTED** [review]
Looks good overall, one thing is that we should make time_col mandatory (even for ordered datasets). Also, maybe adding some quick comments for helper function to remember what they do and how to use them

File: opendg/graph.py (hunk 5 above)
Comment (Jacob-Chmura) [reply opendg/graph.py:52]: Okay, understood. I misunderstood that ordered implies just an enumerated type (with no notion of time). If this _was_ the case, I thought the dataframe _index_ itself provides an implicit order. But I'm happy to make the required.

File: test/test_io/test_csv_io.py (hunk 6 above)
Comment (Jacob-Chmura) [reply test/test_io/test_csv_io.py:54]: This flags an error that we expect to be thrown when a user tries to read a features list with dim=37, but only 3 column names were provided (we expect 37 feature column names).

One option is to try to implicitly use _all_ other columns in the csv file as a feature dimension column. But this is fairly strict and may be annoying to deal with as a user.

Please let me know if you think other behaviour is better here.
